### PR TITLE
storage: mark Replicas as destroyed

### PR DESCRIPTION
**Pull request by petermattis**
Mark Replicas as destroyed in `Replica.Destroy` and do not perform any
further raft operations after destruction.

This addresses the underlying bug in #7386 that was exacerbated by #7355.

The specific sequence of events in #7386 was that a replica was queued
for replication and the test then removed the replica from the
store. Removal of a replica from a store removes it from the various
queues, though that is done asynchronously. The replica queue merrily
tries to process the now destroyed replica and as a first step tries to
acquire the leader lease. This leader lease acquisition is done via a
raft command, but the state underlying the replica has already been
deleted. Boom!

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/7404)

### CODE REVIEW

**Comment by petermattis** [comment]
This needs a test before merging. Wanted to get early eyes as to whether the approach is good.

**Comment by tschottdorf** [comment]
Approach looks good. I assume this fixes at least the test?

File: storage/replica.go [hunk 1]
```diff
@@ -368,6 +374,8 @@ func (r *Replica) Destroy(origDesc roachpb.RangeDescriptor) error {
 	}
 	// Clear the map.
 	r.mu.pendingCmds = map[storagebase.CmdIDKey]*pendingCmd{}
+	r.mu.internalRaftGroup = nil
+	r.mu.destroyed = roachpb.NewRangeNotFoundError(r.RangeID)
```
Comment (tschottdorf) [inline storage/replica.go]: I think for now go with a generic error so we can spot it easily: `roachpb.NewErrorf("replica was garbage collected")`

It should never bubble up to the user. If so, we want to know about it.

Comment (petermattis) [inline storage/replica.go]: `roachpb.NewError` returns a `*roachpb.Error` which does not implement the `error` interface. I used `util.Errorf` instead.

**Comment by petermattis** [comment]
Yes, this fixes `TestReplicaRemovalCampaign` after I revert the revert of eager replication. 

I've added a simple test that verifies `Replica.withRaftGroup` fails after a replica has been removed.

**Comment by tamird** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 377 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-KKu-2wICcOG1cI5NPc0:-KKu-2wJJJpjaHGBAjgR:-1333972167) ([raw file](https://github.com/cockroachdb/cockroach/blob/458332032184a24129c02f1d6119af050626b42c/storage/replica.go#L377)):*
> ```Go
> 	// Clear the map.
> 	r.mu.pendingCmds = map[storagebase.CmdIDKey]*pendingCmd{}
> 	r.mu.internalRaftGroup = nil
> ```

is this needed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-:-KKu-7GYCOJMqsEZFqJb:1417088721)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 377 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-KKu-2wICcOG1cI5NPc0:-KKu-z5NoaQxaVCJA3ZH:-809497353) ([raw file](https://github.com/cockroachdb/cockroach/blob/458332032184a24129c02f1d6119af050626b42c/storage/replica.go#L377)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> No. @tschottdorf suggested it. We really shouldn't be using the `raft.RawNode` after the local raft data has been destroyed. Do you see any harm?
</details>
hard to say - i think interaction with lazy initialization is subtle enough that i'd omit this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 377 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-KKu-2wICcOG1cI5NPc0:-KKu0OPZXpYvGuMpsWGn:1773213210) ([raw file](https://github.com/cockroachdb/cockroach/blob/458332032184a24129c02f1d6119af050626b42c/storage/replica.go#L377)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> hard to say - i think interaction with lazy initialization is subtle enough that i'd omit this.
</details>
We set `internalRaftGroup` to nil elsewhere (e.g. `setReplicaIDLocked`). I think we're both science-dogging this to some degree. @tschottdorf, @bdarnell?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404)*

**Comment by tschottdorf** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[storage/replica.go, line 377 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-KKu-2wICcOG1cI5NPc0:-KKuC9cs0PH7jJLWs1zo:2700476) ([raw file](https://github.com/cockroachdb/cockroach/blob/458332032184a24129c02f1d6119af050626b42c/storage/replica.go#L377)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> We set `internalRaftGroup` to nil elsewhere (e.g. `setReplicaIDLocked`). I think we're both science-dogging this to some degree. @tschottdorf, @bdarnell?
</details>
in `setReplicaIDLocked` it's required because we want a new Raft group to be created. It makes no difference if every caller uses `withRaftGroup{,Locked}` and so they catch the error. But consider the following code:

```go
if r.mu.internalRaftGroup == nil {
   // call withRaftGroup
}
r.mu.internalRaftGroup.blah
```

If you think that code looks unrealistic, well, `r.tick()` is doing something similar and would fall in the trap. So setting to `nil` is a good idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404)*

**Comment by petermattis** [comment]
@tschottdorf LGTY?

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-:-KKu_p7ClC1fRE5DHeND:746441966)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 377 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-KKu-2wICcOG1cI5NPc0:-KKudmVlAds0cNBo6G5a:-436070387) ([raw file](https://github.com/cockroachdb/cockroach/blob/458332032184a24129c02f1d6119af050626b42c/storage/replica.go#L377)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> in `setReplicaIDLocked` it's required because we want a new Raft group to be created. It makes no difference if every caller uses `withRaftGroup{,Locked}` and so they catch the error. But consider the following code:
> 
> ```go
> if r.mu.internalRaftGroup == nil {
>    // call withRaftGroup
> }
> r.mu.internalRaftGroup.blah
> ```
> 
> If you think that code looks unrealistic, well, `r.tick()` is doing something similar and would fall in the trap. So setting to `nil` is a good idea.
</details>
:disappointed: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404)*

**Comment by tschottdorf** [comment]
yes, sorry. lgtm.

On Wed, Jun 22, 2016, 20:27 Tamir Duberstein <notifications@github.com>
wrote:

> Review status: all files reviewed at latest revision, 2 unresolved
> discussions, all commit checks successful.
> ------------------------------
>
> *storage/replica.go, line 377 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7404#-KKu-2wICcOG1cI5NPc0:-KKudmVlAds0cNBo6G5a:-436070387>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/458332032184a24129c02f1d6119af050626b42c/storage/replica.go#L377>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> in setReplicaIDLocked it's required because we want a new Raft group to
> be created. It makes no difference if every caller uses
> withRaftGroup{,Locked} and so they catch the error. But consider the
> following code:
>
> if r.mu.internalRaftGroup == nil {
>    // call withRaftGroup
> }
> r.mu.internalRaftGroup.blah
>
> If you think that code looks unrealistic, well, r.tick() is doing
> something similar and would fall in the trap. So setting to nil is a good
> idea.
>
>
> 😞
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7404>*
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7404#issuecomment-227916650>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135J0y5oTYoYAjoFO1QzY24pqnZOj4ks5qOdLwgaJpZM4I8K94>
> .
>
-- 

-- Tobias

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7404)*

**Pull request merged**